### PR TITLE
Refactor mission manager

### PR DESCRIPTION
**Pull request by patapufate**
This patch:

- Remove unnecessary boilerplate in CMakeList.txt and Package.xml.
- Remove commented/obsolete code
- Add mantainer and license.
- Add Roslint and fix linter errors.
- Change folder structure for header files. Now implemented as "include/<package_name>/header_file.h".

It needs thorough testing, as I don't have the hardware at hand.

### CODE REVIEW

File: mission_manager/include/mission_manager/behavior.h [hunk 1]
```diff
@@ -0,0 +1,416 @@
...
+    ABS_DATETIME
+  }
+  ts_type_t;
+
+  ts_type_t type;
+  boost::posix_time::time_duration duration;
+  boost::posix_time::ptime time;
+};
+
+// class IFactory;
```
Comment (hidmic) [inline mission_manager/include/mission_manager/behavior.h:81]: @patapufate nit: remove?

File: mission_manager/include/mission_manager/behavior.h [hunk 2]
```diff
@@ -0,0 +1,416 @@
...
+
+class BehaviorXMLParam
```
Comment (hidmic) [inline mission_manager/include/mission_manager/behavior.h:83]: @patapufate why removing `namespace` scoping? Declaring anything in the global namespace is almost always a bad idea.

File: mission_manager/include/mission_manager/behavior.h [hunk 3]
```diff
@@ -0,0 +1,416 @@
...
+  }
+
+  bool startBehavior()
+  {
+    std::string xml_tag = getXmlTag();
+
+    switch (getType())
+    {
+      case BEHAVIOR_TYPE_MSG:
+
```
Comment (hidmic) [inline mission_manager/include/mission_manager/behavior.h:258]: @patapufate unnecessary blank line? Same below.

File: mission_manager/include/mission_manager/behaviors/podlog.h [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+ *********************************************************************/
+
+// Original version: Christopher Scianna Christopher.Scianna@us.QinetiQ.com
+
+#ifndef MISSION_MANAGER_BEHAVIORS_PODLOG_H
+#define MISSION_MANAGER_BEHAVIORS_PODLOG_H
+
+#include <string>
+
+#include "../behavior.h"
```
Comment (hidmic) [inline mission_manager/include/mission_manager/behaviors/podlog.h:42]: @patapufate do not use relative paths:
```suggestion
#include "mission_manager/behavior.h"
```

File: mission_manager/src/behaviors/altitude_heading.cpp [hunk 5]
```diff
@@ -34,22 +34,34 @@
...
+        <altitude_heading>
+            <description>
+                00:00:00 - .
+            </description>
+            <when unit="sec">0</when>
+            <timeout unit="sec">50</timeout>
+            <altitude unit="deg">10.0</altitude>
+            <heading unit="deg">180.0</heading>
+            <speed_knots>0.0</speed_knots>
+        </altitude_heading>
```
Comment (hidmic) [inline mission_manager/src/behaviors/altitude_heading.cpp:48]: @patapufate meta: I think this is relevant documentation that should be public. That is, it should be in the corresponding header file.

File: mission_manager/src/behaviors/waypoint.cpp [hunk 6]
```diff
@@ -175,157 +169,101 @@ void WaypointBehavior::latLongtoUTM(double latitude, double longitude, double* p
 #endif
```
Comment (hidmic) [inline mission_manager/src/behaviors/waypoint.cpp:169]: @patapufate hmm this block seems to be compiled out?

File: mission_manager/src/mission.cpp [hunk 7]
```diff
@@ -78,12 +82,15 @@ void Mission::Stop() {
...
-    do {
+    do
+    {
       cur_behavior->stopBehavior();
       cur_behavior = getNextBehavior();
-
-    } while (cur_behavior != NULL);
-  } else if (GetState() == MissionState::ABORTING) {
+    }
+    while (cur_behavior != NULL);
```
Comment (hidmic) [inline mission_manager/src/mission.cpp:90]: @patapufate nit: can't this `while` statement be right next to the closing curly brace ?

**Review by hidmic — COMMENTED** [review]
First pass. Since we're compiling as C++11, we may as well leverage it e.g. use `nullptr` instead of `NULL`, use range-for instead of iterators (where possible), etc.

File: mission_manager/src/mission.cpp (hunk 7 above)
Comment (patapufate) [reply mission_manager/src/mission.cpp:90]: No. Roslint requires "} should be on a line by itself"

File: mission_manager/src/behaviors/waypoint.cpp (hunk 6 above)
Comment (patapufate) [reply mission_manager/src/behaviors/waypoint.cpp:169]: Yes. Good Catch.

File: mission_manager/src/behaviors/altitude_heading.cpp (hunk 5 above)
Comment (patapufate) [reply mission_manager/src/behaviors/altitude_heading.cpp:48]: Done

File: mission_manager/include/mission_manager/behaviors/podlog.h (hunk 4 above)
Comment (patapufate) [reply mission_manager/include/mission_manager/behaviors/podlog.h:42]: Done.

File: mission_manager/include/mission_manager/behavior.h (hunk 3 above)
Comment (patapufate) [reply mission_manager/include/mission_manager/behavior.h:258]: Fixed

File: mission_manager/include/mission_manager/behavior.h (hunk 1 above)
Comment (patapufate) [reply mission_manager/include/mission_manager/behavior.h:81]: Done

File: mission_manager/include/mission_manager/behavior.h (hunk 2 above)
Comment (patapufate) [reply mission_manager/include/mission_manager/behavior.h:83]: Restored

File: mission_manager/CMakeLists.txt [hunk 8]
```diff
@@ -178,50 +78,51 @@ add_executable(${NODE_APP_NAME}
...
+  include/mission_manager/behaviors/attitude_servo.h
+  include/mission_manager/behaviors/depth_heading.h
+  include/mission_manager/behaviors/fixed_rudder.h
+  include/mission_manager/behaviors/payload_command.h
+  include/mission_manager/behaviors/pinger.h
+  include/mission_manager/behaviors/podlog.h
+  include/mission_manager/behaviors/waypoint.h
+)
+
+roslint_add_test()
```
[the remark this answers is not in the document]
Comment (patapufate) [reply mission_manager/CMakeLists.txt]: Done

File: mission_manager/src/mission.cpp (hunk 7 above)
Comment (hidmic) [reply mission_manager/src/mission.cpp:92]: I see. Thanks for the explanation.

File: mission_manager/CMakeLists.txt [hunk 9]
```diff
@@ -231,11 +130,4 @@ install(FILES
...
 
-## Add gtest based cpp test target and link libraries
-# catkin_add_gtest(${PROJECT_NAME}-test test/test_mission_manager.cpp)
-# if(TARGET ${PROJECT_NAME}-test)
-#   target_link_libraries(${PROJECT_NAME}-test ${PROJECT_NAME})
-# endif()
-
-## Add folders to be run by python nosetests
-# catkin_add_nosetests(test)
+roslint_add_test()
```
Comment (hidmic) [inline mission_manager/CMakeLists.txt:133]: @patapufate nit: missing newline!

File: mission_manager/include/mission_manager/behavior.h [hunk 10]
```diff
@@ -0,0 +1,410 @@
...
+    switch (getType())
+    {
+      case BEHAVIOR_TYPE_MSG:
+        // Publish the appropriate behavior message
+        ROS_INFO("Publishing behavior [%s]", m_xml_tag.c_str());
+        publishMsg();
+        break;
+
+      case BEHAVIOR_TYPE_SRV:
+
```
Comment (hidmic) [inline mission_manager/include/mission_manager/behavior.h:169]: @patapufate nit: unnecessary blank lines?

File: mission_manager/include/mission_manager/behavior.h [hunk 11]
```diff
@@ -0,0 +1,410 @@
...
+    TimeStamp when = getWhen();
+
+    switch (when.type)
+    {
+      case TimeStamp::REL_TIME:
+        if (when.duration.total_seconds() == 0) break;  // now
+        boost::this_thread::sleep(when.duration);
+        break;
+
```
Comment (hidmic) [inline mission_manager/include/mission_manager/behavior.h:191]: @patapufate nit: same about unnecessary blank lines.

File: mission_manager/include/mission_manager/behaviors/altitude_heading.h [hunk 12]
```diff
@@ -34,41 +34,41 @@
...
+            <description>
+                00:00:00 - .
+            </description>
+            <when unit="sec">0</when>
+            <timeout unit="sec">50</timeout>
+            <altitude unit="deg">10.0</altitude>
+            <heading unit="deg">180.0</heading>
+            <speed_knots>0.0</speed_knots>
+        </altitude_heading>
+*/
```
Comment (hidmic) [inline mission_manager/include/mission_manager/behaviors/altitude_heading.h:49]: @patapufate I was thinking of putting it right next to the class definition, as Doxygen block, but OK.

File: mission_manager/src/behaviors/attitude_servo.cpp [hunk 13]
```diff
@@ -63,117 +62,97 @@ AttitudeServoBehavior::AttitudeServoBehavior()
...
-  if (f != 0.0) m_roll_tol = (float)f;
+  if (f != 0.0) m_roll_tol = static_cast<float>(f);
   f = 0.0;
   nh.getParam("/mission_manager_node/attitude_servo_pitch_tol", f);
-  if (f != 0.0) m_pitch_tol = (float)f;
+  if (f != 0.0) m_pitch_tol = static_cast<float>(f);
   f = 0.0;
   nh.getParam("/mission_manager_node/attitude_servo_yaw_tol", f);
-  if (f != 0.0) m_yaw_tol = (float)f;
+  if (f != 0.0) m_yaw_tol = static_cast<float>(f);
```
Comment (hidmic) [inline mission_manager/src/behaviors/attitude_servo.cpp:76]: @patapufate meta: I wonder why not simply using `double` members everywhere.

**Review by hidmic — APPROVED** [review]
LGTM but for a few more comments.

File: mission_manager/CMakeLists.txt (hunk 9 above)
Comment (patapufate) [reply mission_manager/CMakeLists.txt]: Fixed!

File: mission_manager/include/mission_manager/behavior.h (hunk 10 above)
Comment (patapufate) [reply mission_manager/include/mission_manager/behavior.h]: Fixed

File: mission_manager/include/mission_manager/behavior.h (hunk 11 above)
Comment (patapufate) [reply mission_manager/include/mission_manager/behavior.h]: Fixed

**Pull request merged**